lib/pry/deprecate.rb

```diff
@@ -0,0 +1,190 @@
+# coding: utf-8
+module Pry::Deprecate
+  require 'set'
+  require 'monitor'
+  DEPRECATE_PRY_SET = SortedSet.new []
+  DEPRECATE_LOCK = Monitor.new
+  DEPRECATE_QUOTED_FILE = Regexp.quote(__FILE__).freeze
+
+  #
+  # Deprecates a method so that all future calls to it display a deprecation message,
+  # unless _pry_.config.print_deprecations is not 'true'.
+  #
+  # @example
+  #
+  #  deprecate_method! [Kernel.method(:puts), "StringIO#puts"],
+  #                    "#puts is deprecated, use #write instead"
+  #  Kernel.puts "foo"
+  #  StringIO.new("").puts "bar"
+  #
+  # @param [Array<String, Method>] sigs
+  #   An array of signatures in the form of a String as 'Foo::Bar#baz' (instance method),
+  #   'Foo::Bar.baz' (class method) or Method objects.
+  #
+  # @param [Pry] pry
+  #   An instance of Pry. `Pry#output` is used for writing.
+  #   Optional, since defaults to '_pry_' in scope of current 'self'.
+  #
+  # @return [nil]
+  #
+  def deprecate_method! sigs, message, pry=((defined?(_pry_) and _pry_) or raise)
+    DEPRECATE_LOCK.synchronize &proc {
```

Comment:
We mentioned this in IRC but mentioning it here just in case:

can just do `synchronize {}`, don't need the `&proc {}`